File: nmpolicy/internal/resolver/replace.go
/*
 * Copyright 2021 NMPolicy Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 *	  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package resolver

import (
	"github.com/nmstate/nmpolicy/nmpolicy/internal/ast"
)

func replace(inputState map[string]interface{}, pathSteps ast.VariadicOperator, replaceValue interface{}) (map[string]interface{}, error) {
	pathVisitorWithReplaceOp := pathVisitor{
		stateVisitor: &replaceOpVisitor{
			replaceValue: replaceValue,
		},
	}
	replaced, err := pathVisitorWithReplaceOp.visitNextStep(path{steps: pathSteps}, inputState)

	if err != nil {
		return nil, replaceError("failed applying operation on the path: %w", err)
	}

	replacedMap, ok := replaced.(map[string]interface{})
	if !ok {
		return nil, replaceError("failed converting result to a map")
	}
	return replacedMap, nil
}

type replaceOpVisitor struct {
	replaceValue interface{}
}

func (r *replaceOpVisitor) visitLastMap(inputMap map[string]interface{}, mapEntryKeyToReplace string) (interface{}, error) {
	modifiedMap := map[string]interface{}{}
	for k, v := range inputMap {
		modifiedMap[k] = v
	}

	modifiedMap[mapEntryKeyToReplace] = r.replaceValue
	return modifiedMap, nil
}

func (*replaceOpVisitor) visitLastSlice([]interface{}, int) (interface{}, error) {
	return nil, nil
}

func (*replaceOpVisitor) visitMapWithIdentity(sv stepVisitor, p path,
	mapToVisit map[string]interface{}, identity string) (interface{}, error) {
	interfaceToVisit, ok := mapToVisit[identity]
	if !ok {
		return nil, nil
	}

	visitResult, err := sv.visitNextStep(p, interfaceToVisit)
	if err != nil {
		return nil, err
	}

	replacedMap := map[string]interface{}{}
	for k, v := range mapToVisit {
		replacedMap[k] = v
	}
	replacedMap[identity] = visitResult
	return replacedMap, nil
}

func (r *replaceOpVisitor) visitSliceWithoutIndex(sv stepVisitor, p path, sliceToVisit []interface{}) (interface{}, error) {
	replacedSlice := make([]interface{}, len(sliceToVisit))
	for i, interfaceToVisit := range sliceToVisit {
		visitResult, err := sv.visitNextStep(p, interfaceToVisit)
		if err != nil {
			return nil, err
		}
		replacedSlice[i] = visitResult
	}
	return replacedSlice, nil
}

func (*replaceOpVisitor) visitSliceWithIndex(stepVisitor, path, []interface{}, int) (interface{}, error) {
	return nil, nil
}

File: rust/src/lib/resolve/replace.rs
use crate::{
    error::{evaluation_error, ErrorKind, NmpolicyError},
    resolve::{
        path::{Path, Step},
        visitor,
    },
    types::NMState,
};

use serde_json::{Map, Value};

struct ReplaceVisitor {
    replace_value: Value,
}

pub(crate) fn visit_state(
    input_state: NMState,
    path: Path,
    replace_value: Value,
) -> Result<NMState, NmpolicyError> {
    match visitor::visit_state(
        path,
        Value::Object(input_state),
        &mut ReplaceVisitor { replace_value },
    ) {
        Ok(replaced) => match replaced {
            Value::Object(map) => Ok(map),
            Value::Null => Ok(NMState::new()),
            _ => Err(evaluation_error(
                "failed converting result to a map".to_string(),
            )),
        },
        Err(e) => Err(e.ctx("failed applying operation on the path".to_string())),
    }
}

impl visitor::StateVisitor for ReplaceVisitor {
    fn visit_last_map(
        &mut self,
        path: Path,
        mut map_to_visit: Map<String, Value>,
    ) -> Result<Value, NmpolicyError> {
        match path.current_step() {
            Step::Identity(_, step) => {
                map_to_visit.insert(step.clone(), self.replace_value.clone());
                Ok(Value::from(map_to_visit))
            }
            _ => Err(NmpolicyError::new(ErrorKind::NotImplementedError)),
        }
    }

    fn visit_last_slice(
        &mut self,
        path: Path,
        slice_to_visit: Vec<Value>,
    ) -> Result<Value, NmpolicyError> {
        match path.current_step() {
            Step::Identity(_, _) => self.visit_slice(path, slice_to_visit),
            _ => Err(NmpolicyError::new(ErrorKind::NotImplementedError)),
        }
    }

    fn visit_map(
        &mut self,
        mut path: Path,
        mut map_to_visit: Map<String, Value>,
    ) -> Result<Value, NmpolicyError> {
        match path.clone().current_step() {
            Step::Identity(_, step) => match map_to_visit.get(step) {
                Some(value_to_visit) => {
                    path.next_step();
                    let visit_result = visitor::visit_state(path, value_to_visit.clone(), self)?;
                    map_to_visit.insert(step.clone(), visit_result);
                    Ok(Value::from(map_to_visit))
                }
                None => Ok(Value::Null),
            },
            _ => Err(NmpolicyError::new(ErrorKind::NotSupportedError)),
        }
    }

    fn visit_slice(
        &mut self,
        path: Path,
        slice_to_visit: Vec<Value>,
    ) -> Result<Value, NmpolicyError> {
        match path.current_step() {
            Step::Identity(_, _) => {
                let result: Result<Vec<Value>, NmpolicyError> = slice_to_visit
                    .into_iter()
                    .map(|v| visitor::visit_state(path.clone(), v, self))
                    .collect();
                Ok(Value::from(result?))
            }
            _ => Err(NmpolicyError::new(ErrorKind::NotSupportedError)),
        }
    }
}
